Fix plugin macros not being exposed through airflow.macros

RikHeijdens opened the pull request:
This PR fixes an issue where macros that are being provided through plugins can not be used at template time because they are not accessible through the `airflow.macros` module.

This PR consists out of two commits. In the first commit I add a test-case that reproduces the issue as outlined in #12785, and the second commit introduces the fix which fixes the issue and allows the test-case to pass.

Fixes: #12785 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

Commit:
Add a test that reproduces airflow#12785

Added a test case to reproduce the issue reported in
https://github.com/apache/airflow/issues/12785

Commit:
Fix plugin-provided macros not being exposed on airflow.macros

In order to allow a plugin-provided macro to be used at templating time,
it needs to be exposed through the airflow.macros module.

This commit fixes GitHub issue airflow#12785.

RikHeijdens commented on tests/plugins/test_plugins_manager.py:
```diff
@@ -213,6 +214,34 @@ def test_entrypoint_plugin_errors_dont_raise_exceptions(self, caplog):
+            # Verify that the symbol table in airflow.macros has been updated with an entry for
+            # this plugin, this is necessary in order to allow the plugin's macros to be used when
+            # rendering templates.
+            assert hasattr(macros, MacroPlugin.name)
```
`integrate_macros_plugins()` has the side-effect that it modifies the `airflow.macros` module. Do we need to add additional cleanup logic to reset the contents of `airflow.macros` through a finalizer here?

mik-laj replied:
Will [``importlib.reload``](https://docs.python.org/3/library/importlib.html#importlib.reload) help?

RikHeijdens replied:
Simply reloading the module using `importlib.reload()` will not be sufficient as the module's symbol table (dictionary) is [retained](https://docs.python.org/3/library/importlib.html#importlib.reload).

> When a module is reloaded, its dictionary (containing the module’s global variables) is retained. Redefinitions of names will override the old definitions, so this is generally not a problem. If the new version of a module does not define a name that was defined by the old version, the old definition remains.

I think we'll have to delete the module from `sys.modules` and import it again in order to properly remove the entries that are being added by this test case.

RikHeijdens replied:
I've implemented this in https://github.com/apache/airflow/pull/12788/commits/c2b5354addb6051f689061eeb9d7ea4d6d8965e3 -- I'm curious to your thoughts.

mik-laj replied:
I love this change.😻  I'm just wondering if it's worth moving this code to the context manager mock_plugins_manager. This allows us to limit the side effects in other tests as well. What do you think?

RikHeijdens replied:
I considered doing this, but the `__doc__` comment on the `mock_plugin_manager` seems to suggest that the scope of that mock is limited to the `airflow.plugins` module.

While that mock does in fact clear out the `macros_modules` variable in `airflow.plugins`, it does not actually attempt to reverse any (side) effects that are caused by invoking `integrate_macros_plugins()`. I did a bit of searching through the code base, and it doesn't really appear that there is any test coverage for this function beyond the test I've just added.

Because I want to avoid scope creep for the `mock_plugin_manager` fixture, and this is the only test case that actually appears to have to deal with side effects cause by calling `integrate_macros_plugins()`, I'm a bit hesitant to make changes beyond what's being proposed here.

mik-laj commented on airflow/plugins_manager.py:
```diff
@@ -419,4 +420,9 @@ def integrate_macros_plugins() -> None:
+            sys.modules[macros_module.__name__] = macros_module
+            # Register the newly created module on airflow.macros such that it
+            # can be accessed when rendering templates.
+            setattr(macros, macros_module.__name__.split('.')[-1], macros_module)
```
```suggestion
            setattr(macros, plugin.name, macros_module)
```

RikHeijdens replied:
Fixed: https://github.com/apache/airflow/pull/12788/commits/84dc19e62f31ae1722005793029df53738f03eed

Commit:
Use plugin.name instead of splitting the module name

Commit:
Add cleanup logic to test_registering_plugin_macros

This test-case has side-effects in the sense that the symbol table of
the airflow.macros module is altered when integrate_macros_plugins() is
invoked. This commit adds a finalizer to the test case that ensures that
that module is being reloaded completely in order to prevent impact on
other tests.

Commit:
Remove access to airflow.macros for subprocesses

The process that runs in the virtual environment may not be able to
access macros that have been provided by plugins. E.g. in cases where
the virtual environment uses a different major version of Python than
Airflow itself.

In order to prevent this from being an issue macros has been removed
from the serializable context.

Commit:
Revert "Remove access to airflow.macros for subprocesses"

This reverts commit 5a6c3cfd19c584a2b5ab2220ee580f8fd85ddb17.

Commit:
Integrate plugin-provided macros in subprocesses

When Airflow is available in a virtual environment, and when this
environment runs at least Python 3, then plugin-provided macros should
be made available to the Python callable that is being executed in this
environment.

mik-laj commented on airflow/utils/python_virtualenv_script.jinja2:
```diff
@@ -20,6 +20,18 @@
+# Check whether Airflow is available in the environment.
+# If it is, we'll want to ensure that we integrate any macros that are being provided
+# by plugins prior to unpickling the task context.
+if int(sys.version[0]) >= 3:
```
Airflow 2.0 only supports Python 3.6 and up. https://github.com/apache/airflow#requirements

RikHeijdens replied:
I'll make the check a bit more strict, just in case :-)

Edit: This has been addressed in https://github.com/apache/airflow/pull/12788/commits/cee7686ab15b5acdc4967ed775b00af48cd5990c.

Commit:
Implement a more strict version-check

Commit:
Document macros limitation

Plugin-provided macros can not be used on Python 2 when using
PythonVirtualenvOperator any longer.
